flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/DefaultStreamGraphContext.java

```diff
@@ -0,0 +1,222 @@
+            LOG.info(
+                    "Modification for edge {} is not allowed as the origin partitioner is ForwardPartitioner.",
+                    targetEdge);
+            return false;
+        }
+
+        return true;
+    }
+
+    private void modifyOutputPartitioner(
+            StreamEdge targetEdge, StreamPartitioner<?> newPartitioner) {
+        if (newPartitioner == null || targetEdge == null) {
+            return;
+        }
+        Integer sourceNodeId = targetEdge.getSourceId();
+        Integer targetNodeId = targetEdge.getTargetId();
+
+        StreamPartitioner<?> oldPartitioner = targetEdge.getPartitioner();
+
+        targetEdge.setPartitioner(newPartitioner);
+
+        // For non-chainable edges, we change the ForwardPartitioner to RescalePartitioner to avoid
+        // limiting the parallelism of the downstream node by the forward edge.
+        // 1. If the upstream job vertex is created.
+        if (targetEdge.getPartitioner() instanceof ForwardPartitioner
+                && frozenNodeToStartNodeMap.containsKey(sourceNodeId)) {
+            targetEdge.setPartitioner(new RescalePartitioner<>());
+        }
+        // 2. If the source and target are non-chainable.
+        if (targetEdge.getPartitioner() instanceof ForwardPartitioner
+                && !StreamingJobGraphGenerator.isChainable(targetEdge, streamGraph)) {
+            targetEdge.setPartitioner(new RescalePartitioner<>());
+        }
+        // 3. If the forward group cannot be merged.
+        if (targetEdge.getPartitioner() instanceof ForwardPartitioner
+                && !mergeForwardGroups(sourceNodeId, targetNodeId)) {
+            targetEdge.setPartitioner(new RescalePartitioner<>());
+        }
+
+        Map<StreamEdge, NonChainedOutput> opIntermediateOutputs =
+                opIntermediateOutputsCaches.get(sourceNodeId);
+        NonChainedOutput output =
+                opIntermediateOutputs != null ? opIntermediateOutputs.get(targetEdge) : null;
+        if (output != null) {
+            output.setPartitioner(targetEdge.getPartitioner());
+        }
+        LOG.info(
+                "The original partitioner of the edge {} is: {} , requested change to: {} , and finally modified to: {}.",
+                targetEdge,
+                oldPartitioner,
+                newPartitioner,
+                targetEdge.getPartitioner());
+    }
+
+    private boolean mergeForwardGroups(Integer sourceNodeId, Integer targetNodeId) {
+        StreamNodeForwardGroup sourceForwardGroup =
+                startAndEndNodeIdToForwardGroupMap.get(sourceNodeId);
+        StreamNodeForwardGroup targetForwardGroup =
+                startAndEndNodeIdToForwardGroupMap.get(targetNodeId);
+        if (sourceForwardGroup == null || targetForwardGroup == null) {
+            return false;
+        }
+
+        if (!ForwardGroupComputeUtil.canTargetMergeIntoSourceForwardGroup(
```

Comment:
Looks to it it invokes `canTargetMergeIntoSourceForwardGroup(...)` twice. One here and one in `mergeForwardGroup(...)`.